lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,288 @@
+                val error = getHydrationInputErrors(
+                        unwrappedHydrationSourceFieldType.unwrapOne(),
+                        unwrappedActorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+                if (error != null) {
+                    return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                            parent,
+                            overallField,
+                            remoteArg,
+                            hydrationSourceFieldType,
+                            actorFieldArgType,
+                            actorFieldName
+                    )
+                }
+
+            }
+            // Otherwise we can just check the types normally
+            else {
+                return getHydrationInputErrors(
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydration,
+                        actorFieldName
+                )
+            }
+            return null
+        }
+
+        private fun getHydrationInputErrors(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration,
+                actorFieldName: String
+        ): NadelSchemaValidationError? {
+
+            //need to check null compatibility
+            val sourceType = remoteArg.remoteArgumentSource.sourceType
+            if (sourceType != ObjectField && actorFieldArgType.isNonNull && !hydrationSourceFieldType.isNonNull) {
+                // source must be at least as strict as field argument
+                return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                        actorFieldName
+                )
+            }
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull()
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull()
+
+            // scalar feed into scalar
+            if (unwrappedHydrationSourceFieldType is GraphQLScalarType && unwrappedActorFieldArgType is GraphQLScalarType) {
+                return validateScalarArg(hydrationSourceFieldType, actorFieldArgType, parent, overallField, remoteArg, actorFieldName)
```

Comment:
Same comment as before, can you pass in `unwrappedHydrationSourceFieldType` here instead of passing in the raw `GraphQLType` and then unwrapping it inside the function?

Reply:
passing in wrapped type because we want to preserve the wrapped type for the error message to avoid any confusion from the user

Reply:
the only reason we could pass in the unwrapped type previously (in the case that its an object) is because we do not show the type in that one error message